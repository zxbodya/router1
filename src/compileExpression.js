//todo: tests for expression language
const compileExpression = (expression)=> {
  //todo: for non-production verify expression syntax
  //todo: if syntax is not valid show messages about typical errors

  // http://stackoverflow.com/questions/8844256/split-string-including-regular-expression-match
  let textParts = expression.split(/<([^>]+?)(?::([^>]+))?>/);
  let matcher = [];
  let generateParts = [];
  let paramNames = [];
  for (let i = 0, l = textParts.length; i < l; i++) {
    let part = textParts[i];
    if (i % 3 === 0) {
      //  http://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex#answer-6969486
      matcher.push(part.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&"));
      generateParts.push(part);
    }
    if (i % 3 === 1) {
      paramNames.push(part);
      generateParts.push(null);
    }
    if (i % 3 === 2) {
      //todo: verify expression if it has captures - throw an error
      matcher.push('(', part ? part : '.*', ')');
    }
  }
  return [matcher.join(''), generateParts, paramNames]
};

module.exports = compileExpression;
