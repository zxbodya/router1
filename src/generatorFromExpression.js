let generatorFromExpression = (info)=> {
  let generateParts = info[1], paramNames = info[2];

  return (params = {}) => {
    if (process.env.NODE_ENV !== 'production') {
      var missingParams = paramNames.filter(paramName=>!(paramName in params));

      if (missingParams.length) {
        throw new Error(`missing parameters [${missingParams.join(',')}]`);
      }
    }

    let res = [];
    for (let i = 0, l = generateParts.length, pn = 0, pl = paramNames.length; i < l; i++) {
      let g = generateParts[i];
      if (g !== null) {
        res.push(g);
      } else {
        if (pn < pl) {
          res.push(params[paramNames[pn]]);
          pn++;
        }
      }
    }
    //todo: test if matcher can match the result, and throw if it can not
    return res.join('');
  };
};

export default generatorFromExpression;
