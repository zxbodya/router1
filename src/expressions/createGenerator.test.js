import compileExpression from './compile.js';
import generatorFromExpression from './createGenerator.js';

describe('Router, generator from expression', function () {
  it('it generates strings without params', ()=> {
    let expression = compileExpression('aaaaa');
    let generator = generatorFromExpression(expression);

    expect(generator()).toBe('aaaaa');
    expect(generator({})).toBe('aaaaa');
    expect(generator({a: 123, b: 321})).toBe('aaaaa');
  });

  it('it generates strings with one param', ()=> {
    let expression = compileExpression('aaaaa<a>');
    let generator = generatorFromExpression(expression);

    expect(generator({a: 123})).toBe('aaaaa123');
    expect(generator({a: 321})).toBe('aaaaa321');
    expect(generator({a: 321, b: 444})).toBe('aaaaa321');
  });

  it('it generates strings with few params', ()=> {
    let expression = compileExpression('aaaaa<a>-<b>');
    let generator = generatorFromExpression(expression);

    expect(generator({a: 321, b: 1})).toBe('aaaaa321-1');
    expect(generator({a: 321, b: 444})).toBe('aaaaa321-444');
  });

  it('throws when parameters are missing', ()=> {
    let expression = compileExpression('aaaaa<a>-<b>');
    let generator = generatorFromExpression(expression);

    expect(()=>generator({a: 321, b: 1})).not.toThrow();
    expect(()=>generator({a: 321})).toThrow();
    expect(()=>generator({b: 1})).toThrow();
    expect(()=>generator({})).toThrow();
  });
});
