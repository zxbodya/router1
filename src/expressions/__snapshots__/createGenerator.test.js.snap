// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router, generator from expression encodes uri 1`] = `"%D0%BF%D1%80%D0%B8%D0%B2%D1%96%D1%82/%3F%23aZ09.-_~!$&%5C'()*+,;=:@"`;
